File: client/transfer_money.py
class TransferRequest:
    def __init__(self, is_deposit: bool, amount: int, acc_no: int, acc_no_dst: int,
                 name: str, pwd: str, currency: str):
        self.is_deposit = is_deposit
        self.amount = amount
        self.acc_no = acc_no
        self.acc_no_dst = acc_no_dst
        self.name = name
        self.pwd = pwd
        self.currency = currency

    def marshal(self):
        deposit = 1 if self.is_deposit else 0
        serialized = bytearray(deposit.to_bytes(1, 'big'))
        serialized.extend(self.amount.to_bytes(8, 'big'))
        serialized.extend(self.acc_no.to_bytes(8, 'big'))
        serialized.extend(self.acc_no_dst.to_bytes(8, 'big'))

        name_size = len(self.name)
        pwd_size = len(self.pwd)
        currency_size = len(self.currency)

        serialized.extend(name_size.to_bytes(2, 'big'))
        serialized.extend(pwd_size.to_bytes(2, 'big'))
        serialized.extend(currency_size.to_bytes(2, 'big'))

        serialized.extend(bytearray(self.name, 'utf-8'))
        serialized.extend(bytearray(self.pwd, 'utf-8'))
        serialized.extend(bytearray(self.currency, 'utf-8'))
        return serialized


class TransferResponse:
    def __init__(self, balance: int):
        self.balance = balance

    def __str__(self):
        return f"Your updated balance is {self.balance}"

    @classmethod
    def unmarshal(cls, data) -> str:
        balance = int.from_bytes(data[:8], 'big')
        return str(TransferResponse(balance))


class TransferMonitorResponse:
    def __init__(self, acc_no: int, acc_no_dst: int, amount: int):
        self.acc_no = acc_no
        self.acc_no_dst = acc_no_dst
        self.amount = amount

    def __str__(self):
        return f"Amount {self.amount} is transferred from Account number {self.acc_no} " \
               f"to number Account {self.acc_no_dst}"

    @classmethod
    def unmarshal(cls, data: bytes) -> str:
        acc_no = int.from_bytes(data[0:8], 'big')
        acc_no_dst = int.from_bytes(data[8:16], 'big')
        amount = int.from_bytes(data[16:24], 'big')
        return str(TransferMonitorResponse(
            acc_no=acc_no,
            acc_no_dst=acc_no_dst,
            amount=amount
        ))

File: account/transfer_money.go
package account

import (
	"encoding/binary"
	"fmt"
	"math"
)

type transferRequest struct {
	amount       float64
	accNumber    uint64
	accNumberDst uint64
	name         string
	password     string
	currency     string
}

func (req *transferRequest) unmarshal(data []byte) {
	req.amount = math.Float64frombits(binary.BigEndian.Uint64(data[0:8]))
	req.accNumber = binary.BigEndian.Uint64(data[8:16])
	req.accNumberDst = binary.BigEndian.Uint64(data[16:24])

	nameSize := binary.BigEndian.Uint16(data[24:26])
	pwdSize := binary.BigEndian.Uint16(data[26:28])
	currencySize := binary.BigEndian.Uint16(data[28:30])

	pwdIndex := 30 + nameSize
	req.name = string(data[30:pwdIndex])

	currencyIndex := pwdIndex + pwdSize
	req.password = string(data[pwdIndex:currencyIndex])

	req.currency = string(data[currencyIndex : currencyIndex+currencySize])
}

type transferResponse struct {
	balance float64
}

func (res *transferResponse) marshal() []byte {
	arr := make([]byte, 8)
	binary.BigEndian.PutUint64(arr, math.Float64bits(res.balance))
	return arr
}

func TransferMoney(content []byte) (StatusCode, []byte) {
	req := &transferRequest{}
	req.unmarshal(content)

	// Validation
	authCode := Database.authenticate(req.accNumber, req.name, req.password)
	if authCode != SUCCESS {
		return authCode, nil
	}

	account := Database.records[req.accNumber]
	accountDst, ok := Database.records[req.accNumberDst]
	if !ok {
		return INVALID_RECIPIENT_ACCOUNT, nil
	}

	if req.currency != account.Currency {
		return WRONG_CURRENCY, nil
	}
	if req.currency != accountDst.Currency {
		return WRONG_RECIPIENT_CURRENCY, nil
	}

	if account.Balance < req.amount {
		return INSUFFICIENT_BALANCE, nil
	}

	account.Balance -= req.amount
	accountDst.Balance += req.amount

	// Prepare monitor dispatch
	s := fmt.Sprintf("Amount %f is transferred from Account number %d to Account number %d",
		req.amount, req.accNumber, req.accNumberDst)
	clientsTrackingImpl.dispatchEvent([]byte(s))

	// Prepare response
	res := &transferResponse{balance: account.Balance}
	return SUCCESS, res.marshal()
}
